src/EFCore.PG/Storage/Internal/NpgsqlRelationalConnection.cs

```diff
@@ -35,8 +40,14 @@ protected override DbConnection CreateDbConnection()
                 conn.ProvideClientCertificatesCallback = ProvideClientCertificatesCallback;
             if (RemoteCertificateValidationCallback != null)
                 conn.UserCertificateValidationCallback = RemoteCertificateValidationCallback;
+
+            if (_validated)
```

Comment:
I don't think this will actually work - RelationalConnection has a DI service lifetime of "scoped", meaning that each DbContext has its own instance.

We can make this flag static and specific to the MaxAutoPrepare check (merging this check with the one below). It's not ideal to use statics with dependency injection but I don't think there's an appropriate singleton service for doing this check.

Reply:
That's a good point... But a simple static flag also seems insufficient in the case of two or more connection strings (e.g. second one should warn, but the first already set the flag). 

Maybe a `Dictionary<NpgsqlConnectionStringBuilder, bool>` instead?

On the other hand, thinking about this in terms of a web app (e.g. `scope == request`), is one warning per-request really too noisy (i.e. given that consumers can always disable this warning if its by design)?

Reply:
I don't like the idea with a static variable because EF allows to register all required stuff in any container, but by default it uses a static service provider cache. So the best way is to design a log settings manager and register it inside `InternalServiceProvider`. It seems that some gymnastic is required here, but things are much simpler. Just register the manager inside `AddEntityFrameworkNpgsql`. The manager should store the variable you invented.

Reply:
Adding a dictionary lookup on every connection open just to issue a warning seems like a perf no-no to me.

One global to say that at least one connection has been made without autoprepare seems sufficient to me - no real need to do so on a connstring-by-connstring basis. IMHO the idea is to globally bring it to the user's attention.

@YohDeadfall it's true that a static flag goes outside the DI framework, but I can't see an actual problem here - it's only used to make sure the warning is issued once. I'd rather us not start architecturing complicated stuff just for a warning...

Reply:
> Adding a dictionary lookup on every connection open just to issue a warning seems like a perf no-no to me.

> One global to say that at least one connection has been made without autoprepare seems sufficient to me 

> I'd rather us not start architecturing complicated stuff just for a warning...

I'll rework it to show this with the static flag. Definitely don't want to overcomplicate the hot path with "nice to have" warnings.

> So the best way is to design a log settings manager 

Having a good place to hook-in for validation warnings would be nice, but that feels like something that needs to be done upstream. Might be worth opening an issue?

Reply:
It's up to you, but the implementation for now is very simple (one bool field inside the manager).

Reply:
> Having a good place to hook-in for validation warnings would be nice, but that feels like something that needs to be done upstream. Might be worth opening an issue?

I don't remember hearing about any need for validating a connection string anywhere. In fact, this specifically is really validation as there's nothing wrong with using EF Core without autopreparation - it'll just be slower and we want the user to know about it. To me this scenario seems to specific for a general extension point...